### PR TITLE
CHOMP: General fixup

### PR DESCRIPTION
**Pull request by bmagyar**
### Description

- [x] Fix & test: CHOMP crashes when the planning fails. Currently result fields are copied regardless of the planning success which leads to this behaviour. 
- [x] Fix & test: CHOMP doesn't check for existing start & goal states
- [x] Fix & test: CHOMP doesn't check for joint limits in start & goal states
- [x] Fix & test: CHOMP only supports joint position goals, should reject Cartesian ones
- [x] Fix & test: CHOMP doesn't check end of trajectory within tolerances, _fix done_
- [x] Fix & test: CHOMP path is not checked for collisions

This changes a lot, I don't think we should backport it to Indigo unless directly asked for.

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [x] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

### CODE REVIEW

**Comment by v4hn** [comment]
format checking failed.
I'll wait until after WMD to merge this together with the appropriate test.

**Comment by simonschmeisser** [comment]
Last time I looked there was no check if the final path is actually collision free ... so I guess it is very rare (only due to kinematic limits) that a plan actually fails at all. Just my 2 cents if somebody should be wondering why this didn't crash more often.

**Comment by davetcoleman** [comment]
lgmt but i agree a test would make a great addition to this PR

**Comment by bmagyar** [comment]
The reason I didn't submit the test yesterday was that I was simply unable to reproduce a failure... 

CHOMP at the moment seems to be perfectly fine with both collisions and not reaching goals...
This PR has suddenly became a bit broader :)

**Comment by bmagyar** [comment]
I've updated the description with what I've done so far and a few todos. Could I please get a bit of general feedback @davetcoleman @v4hn ?

**Comment by davetcoleman** [comment]
>  Fix & test: CHOMP only supports position goals, should reject Cartesian ones

When you say "position goals" do you mean joint position goals rather than end effector position goals? Just wanted to clarify.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp [hunk 1]
```diff
@@ -95,6 +117,15 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
     }
   }
 
+  const Eigen::MatrixXd goal_state = trajectory.getTrajectoryPoint(goal_index);
```
Comment (davetcoleman) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: why ``Xd`` instead of Affine3d like we normally use in moveit?

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp [hunk 2]
```diff
@@ -52,16 +52,38 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
...
+  }
+
+  if (req.start_state.joint_state.position.empty())
+  {
+    ROS_ERROR_STREAM("Starting state is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  if (not planning_scene->getRobotModel()->satisfiesPositionBounds(req.start_state.joint_state.position.data()))
```
[the remark this answers is not in the document]
Comment (simonschmeisser) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: this is actually older than c++11 but microsoft decided not to support it because it would break compatibility with C code were it could have been defined using macros. Besides who wants to follow standards anyway?

Theres also 'and', 'or', ... turning it into a slightly less ugly language. See http://en.cppreference.com/w/cpp/language/operator_alternative

[the remark this answers is not in the document]
Comment (bmagyar) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: I find them cute in their own weird way :)

**Comment by bmagyar** [comment]
It is joint position goals, I've updated the description

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp (hunk 1 above)
Comment (bmagyar) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: That's what the `ChompTrajectory` API uses.

File: moveit_planners/chomp/chomp_interface/test/chomp_moveit_test.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+
```
[the remark this answers is not in the document]
Comment (bmagyar) [reply moveit_planners/chomp/chomp_interface/test/chomp_moveit_test.cpp]: Added

File: moveit_planners/chomp/chomp_interface/test/rrbot.srdf [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" ?>
```
[the remark this answers is not in the document]
Comment (bmagyar) [reply moveit_planners/chomp/chomp_interface/test/rrbot.srdf]: I personally prefer to have packages carry their own resources for unit tests.

Could still move them, just say the word ;)

**Comment by bmagyar** [comment]
I have the test for the goal tolerance violation but couldn't get CHOMP to get a suboptimal result yet. Almost there! :) 

@davetcoleman could you quickly check the rest of the changes please?

Note: there are a big bunch of ROS_INFO-s within other parts of the planner, I could remove those as well, let me know.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp [hunk 5]
```diff
@@ -138,13 +145,13 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
...
   {
-    ROS_WARN_STREAM("Could not initialize optimizer");
+    ROS_ERROR_STREAM_NAMED("chomp_planner", "Could not initialize optimizer");
     res.error_code.val = moveit_msgs::MoveItErrorCodes::PLANNING_FAILED;
     return false;
   }
   ROS_DEBUG("Optimization took %f sec to create", (ros::WallTime::now() - create_time).toSec());
   optimizer.optimize();
-  ROS_INFO("Optimization actually took %f sec to run", (ros::WallTime::now() - create_time).toSec());
+  ROS_DEBUG("Optimization actually took %f sec to run", (ros::WallTime::now() - create_time).toSec());
```
Comment (davetcoleman) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: ROS_DEBUG_NAMED("chomp_planner",

Comment (bmagyar) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: I'll update the debug statements to named as well, good catch!

**Comment by bmagyar** [comment]
The tests are ok, they were failing on clang which I fixedin the last commit (not really visible but check the bottom of the PR for latest.

**Comment by bmagyar** [comment]
I've also added CHOMP to the planners released with `moveit_planners`

**Comment by davetcoleman** [comment]
> I've also added CHOMP to the planners released with moveit_planners

Lets leave that change for a separate PR once we confirm the many changes in this PR are good. There will need to be additional steps that @130s should know about

**Comment by bmagyar** [comment]
:+1: I'll open a new PR with that bit

**Comment by bmagyar** [comment]
About the last test: I couldn't reproduce precision issues on the goal with successful plans. We could add creating this to the issue tracker.

**Comment by bmagyar** [comment]
@simonschmeisser any comments?

**Comment by simonschmeisser** [comment]
@bmagyar I'm now always getting collisions (which is an improvement from a safety point of view ;) ) Will debug some more. I somehow suspect the 90° rotation our test scene has between world_frame and base_link. Stomp had some issues here as well ( https://github.com/ros-industrial/industrial_moveit/pull/46 ) Testcase in https://github.com/isys-vision/industrial_moveit/commit/f34caae8b516a8c2b4395969305d4a8144755840

**Comment by bmagyar** [comment]
Thanks! This wouldn't change correctness but make things a wee bit faster,  right? (I'm asking with your previous edited answer in mind)

File: moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<launch>
+	<!-- CHOMP Plugin for MoveIt! -->
+	<arg name="planning_plugin" value="chomp_interface/CHOMPPlanner" />
+
+	<arg name="start_state_max_bounds_error" value="0.1" />
+
+	<param name="planning_plugin" value="$(arg planning_plugin)" />
+	<param name="start_state_max_bounds_error" value="$(arg start_state_max_bounds_error)" />
+
+	<param name="collision_detector" value="Hybrid" />
```
Comment (simonschmeisser) [inline moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml]: "Hybrid" is a bit non-descriptive but maybe we can leave that to another PR

File: moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+	<rosparam command="load" file="$(find moveit_planners_chomp)/test/chomp_planning.yaml"/>
```
Comment (simonschmeisser) [inline moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml]: At some point we should probably put these parameters in a namespace, right now its all cluttered into /move_group

File: moveit_planners/chomp/chomp_interface/test/chomp_planning.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+use_pseudo_inverse: false
+pseudo_inverse_ridge_factor: 1e-4
+animate_endeffector: false
+animate_endeffector_segment: "link3"
+joint_update_limit: 0.1
+collision_clearence: 0.2
+collision_threshold: 0.07
+random_jump_amount: 1.0
+use_stochastic_descent: true
+
```
[the remark this answers is not in the document]
Comment (simonschmeisser) [reply moveit_planners/chomp/chomp_interface/test/chomp_planning.yaml]: I wonder if we should comment out/delete parameters that are only ever used by dead code?

**Comment by simonschmeisser** [comment]
passing the group name will make it ignore collisions between objects that are neither part nor child of the group. That would only be less correct if CHOMP could plan for more than one group simultaneously (it cannot right?)

In summary I think after adding that fix it can be merged and any further work done in new PRs

File: moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml (hunk 6 above)
Comment (bmagyar) [reply moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml]: We should add this to a CHOMP issue list and work on it one issue at a time.

File: moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml (hunk 7 above)
Comment (bmagyar) [reply moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml]: We should add this to a CHOMP issue list and work on it one issue at a time.

**Comment by bmagyar** [comment]
@simonschmeisser , I've done the modification you asked for, now collision check is only done for the planning group at hand (which can only be a single one at a time).

I've also rebased the PR. Once the build passes I think we are good to merge!

**Comment by bmagyar** [comment]
@simonschmeisser , could you please switch change your review to approve as well? (As long as you approve of course :))

**Comment by simonschmeisser** [comment]
I did some short testing again and it looks impressive when it succeeds in planning and communicates correctly when it fails. Great job! lgtm!

File: moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml (hunk 6 above)
Comment (simonschmeisser) [reply moveit_planners/chomp/chomp_interface/test/rrbot_chomp_planning_pipeline.launch.xml]: +1 (as well to the point below) This is a huge improvement and should not be held back any longer!

**Comment by bmagyar** [comment]
Thanks! 
Interestingly you are still listed as a reviewer on this PR. 

I'll go ahead with the merge

**Comment by simonschmeisser** [comment]
Probably somebody else should do the merge?

**Comment by bmagyar** [comment]
There are plenty of feature changes that are separated into individual commits with minimal cluttering.  I propose a _merge via merge commit_ to add meaningful info to the changelog

**Comment by bmagyar** [comment]
@simonschmeisser sure,  I nominate @v4hn !

**Comment by davetcoleman** [comment]
Since @simonschmeisser gave this a +1, I'll merge
